scheds/rust/scx_rustlite/src/main.rs

```diff
@@ -0,0 +1,314 @@
+        let struct_ops = Some(
+            skel.maps_mut()
+                .rustlite()
+                .attach_struct_ops()
+                .context("Failed to attach struct ops")?,
+        );
+        info!("{} scheduler attached", SCHEDULER_NAME);
+
+        // Return scheduler object.
+        Ok(Self {
+            skel,
+            task_pool,
+            struct_ops,
+        })
+    }
+
+    fn read_bpf_exit_kind(&mut self) -> i32 {
+        unsafe { std::ptr::read_volatile(&self.skel.bss().exit_kind as *const _) }
+    }
+
+    fn report_bpf_exit_kind(&mut self) -> Result<()> {
+        match self.read_bpf_exit_kind() {
+            0 => Ok(()),
+            etype => {
+                let cstr = unsafe { CStr::from_ptr(self.skel.bss().exit_msg.as_ptr() as *const _) };
+                let msg = cstr
+                    .to_str()
+                    .context("Failed to convert exit msg to string")
+                    .unwrap();
+                info!("BPF exit_kind={} msg={}", etype, msg);
+                Ok(())
+            }
+        }
+    }
+
+    fn schedule(&mut self) {
+        let maps = self.skel.maps();
+        let enqueued = maps.enqueued();
+        let dispatched = maps.dispatched();
+
+        // Drain enqueued list and push all the tasks to the task pool (sorted by vtime).
+        loop {
+            match enqueued.lookup_and_delete(&[]) {
+                Ok(Some(val)) => {
+                    let task = unsafe { &*(val.as_slice().as_ptr() as *const bpf_intf::task_ctx) };
+                    self.task_pool.push(task.pid, task.vtime);
+                }
+                Ok(None) => break,
+                Err(err) => {
+                    warn!("Error: {}", err);
+                    break;
+                }
+            }
+        }
+
+        // Dispatch tasks from the task pool in order (sending them to the BPF part).
+        loop {
+            match self.task_pool.pop() {
+                Some(task) => {
+                    let task_struct: &[u8] = unsafe {
+                        std::slice::from_raw_parts(
+                            &task as *const Task as *const u8,
+                            std::mem::size_of::<Task>(),
+                        )
+                    };
+                    match dispatched.update(&[], &task_struct, libbpf_rs::MapFlags::ANY) {
+                        Ok(_) => {}
+                        Err(_) => {
+                            /*
+                             * Re-add the task to the dispatched list in case of failure and stop
+                             * dispatching.
+                             */
+                            self.task_pool.push(task.pid, task.vtime);
+                            break;
+                        }
+                    }
+                }
+                None => break,
+            }
+        }
+
+        // Yield to avoid using too much CPU from the scheduler itself.
+        thread::yield_now();
```

Comment:
Another nice thing about the ringbuffer maps is that user space can block on epoll_wait() when waiting for more messages from the kernel. This of course works just fine as well given that the kernel will only schedule the task usersched task when necessary